demos/game_of_life/sim.hpp

```diff
@@ -0,0 +1,236 @@
+            } catch (std::exception const& e) {
+              std::cout << "Caught SYCL exception:\n" << e.what() << std::endl;
+            }
+          }
+        }) {
+    // Initialize game grid to empty
+    auto acells = m_game.read()
+                      .cells.get_access<sycl::access::mode::discard_write,
+                                        sycl::access::target::host_buffer>();
+    auto aimg = m_game.read()
+                    .img.get_access<sycl::access::mode::discard_write,
+                                    sycl::access::target::host_buffer>();
+    for (size_t y = 0; y < m_height; y++) {
+      for (size_t x = 0; x < m_width; x++) {
+        acells[sycl::id<2>(x, y)] = CellState::DEAD;
+        aimg[sycl::id<2>(y, x)] = sycl::cl_uchar4(0, 0, 0, 0);
+      }
+    }
+  }
+
+  /// Add a button press (cell spawn) to be processed
+  void add_click(size_t x, size_t y, CellState state) {
+    // Click processing is deferred until update
+    m_clicks.emplace_back(x, y, state);
+  }
+
+  void step();
+
+  /// Calls the provided function with image data
+  template <typename Func>
+  void with_img(Func&& func) {
+    auto acc = m_game.read()
+                   .img.get_access<sycl::access::mode::read,
+                                   sycl::access::target::host_buffer>();
+    func(acc.get_pointer());
+  }
+
+ private:
+  /// Executes an update frame
+  void internal_step() {
+    using sycl::access::mode;
+    using sycl::access::target;
+
+    // Read mouse clicks since last frame and apply them to the game.
+    {
+      // Have to write into read-buffer rather than write-buffer, since it is
+      // the read-buffer
+      // that will be read by the kernel.
+      auto acc = this->m_game.read()
+                     .cells.get_access<mode::write, target::host_buffer>();
+
+      while (!m_clicks.empty()) {
+        auto press = m_clicks.back();
+        m_clicks.pop_back();
+        acc[sycl::id<2>(std::get<0>(press), std::get<1>(press))] =
+            std::get<2>(press);
+      }
+    }
+
+    this->m_q.submit([&](sycl::handler& cgh) {
+      auto r = this->m_game.read().cells.get_access<mode::read>(cgh);
+      auto rv = this->m_game.read().vels.get_access<mode::read>(cgh);
+      auto w = this->m_game.write().cells.get_access<mode::discard_write>(cgh);
+      auto wv = this->m_game.write().vels.get_access<mode::discard_write>(cgh);
+      auto img = this->m_game.write().img.get_access<mode::discard_write>(cgh);
+
+      // These dummy variables have to be made to avoid capturing 'this' in the
+      // kernel
+      auto width = this->m_width;
+      auto height = this->m_height;
+
+      cgh.parallel_for<class gameoflifesimkernel>(
+          // Work on each cell in parallel
+          cl::sycl::range<2>(width, height), [=](cl::sycl::item<2> item) {
+            size_t x = item.get(0);
+            size_t y = item.get(1);
+
+            // Unrolled loop for accessing the 3x3 stencil
```

Comment:
do you really need to unroll the loop? have you profiled this and check there is any benefit?